File: vane-annotations/src/main/java/org/oddlama/vane/annotation/processor/ConfigAndLangProcessor.java
package org.oddlama.vane.annotation.processor;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import javax.annotation.processing.AbstractProcessor;
import javax.annotation.processing.RoundEnvironment;
import javax.annotation.processing.SupportedAnnotationTypes;
import javax.annotation.processing.SupportedSourceVersion;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.Element;
import javax.lang.model.element.TypeElement;
import javax.tools.Diagnostic;

@SupportedAnnotationTypes(
    {
        "org.oddlama.vane.annotation.config.ConfigBoolean",
        "org.oddlama.vane.annotation.config.ConfigDict",
        "org.oddlama.vane.annotation.config.ConfigDouble",
        "org.oddlama.vane.annotation.config.ConfigDoubleList",
        "org.oddlama.vane.annotation.config.ConfigExtendedMaterial",
        "org.oddlama.vane.annotation.config.ConfigInt",
        "org.oddlama.vane.annotation.config.ConfigIntList",
        "org.oddlama.vane.annotation.config.ConfigItemStack",
        "org.oddlama.vane.annotation.config.ConfigLong",
        "org.oddlama.vane.annotation.config.ConfigMaterial",
        "org.oddlama.vane.annotation.config.ConfigMaterialMapMapMap",
        "org.oddlama.vane.annotation.config.ConfigMaterialSet",
        "org.oddlama.vane.annotation.config.ConfigString",
        "org.oddlama.vane.annotation.config.ConfigStringList",
        "org.oddlama.vane.annotation.config.ConfigStringListMap",
        "org.oddlama.vane.annotation.config.ConfigVersion",
        "org.oddlama.vane.annotation.lang.LangMessage",
        "org.oddlama.vane.annotation.lang.LangMessageArray",
        "org.oddlama.vane.annotation.lang.LangVersion",
    }
)
@SupportedSourceVersion(SourceVersion.RELEASE_21)
public class ConfigAndLangProcessor extends AbstractProcessor {

    @Override
    public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
        for (var annotation : annotations) {
            roundEnv.getElementsAnnotatedWith(annotation).forEach(e -> verifyType(annotation, e));
        }

        return true;
    }

    private static final Map<String, String> fieldTypeMapping;

    static {
        Map<String, String> map = new HashMap<>();
        map.put("org.oddlama.vane.annotation.config.ConfigBoolean", "boolean");
        map.put("org.oddlama.vane.annotation.config.ConfigDict", "<any>");
        map.put("org.oddlama.vane.annotation.config.ConfigDouble", "double");
        map.put("org.oddlama.vane.annotation.config.ConfigDoubleList", "java.util.List<java.lang.Double>");
        map.put(
            "org.oddlama.vane.annotation.config.ConfigExtendedMaterial",
            "org.oddlama.vane.core.material.ExtendedMaterial"
        );
        map.put("org.oddlama.vane.annotation.config.ConfigInt", "int");
        map.put("org.oddlama.vane.annotation.config.ConfigIntList", "java.util.List<java.lang.Integer>");
        map.put("org.oddlama.vane.annotation.config.ConfigItemStack", "org.bukkit.inventory.ItemStack");
        map.put("org.oddlama.vane.annotation.config.ConfigLong", "long");
        map.put("org.oddlama.vane.annotation.config.ConfigMaterial", "org.bukkit.Material");
        map.put(
            "org.oddlama.vane.annotation.config.ConfigMaterialMapMapMap",
            "java.util.Map<java.lang.String,java.util.Map<java.lang.String,java.util.Map<java.lang.String,org.bukkit.Material>>>"
        );
        map.put("org.oddlama.vane.annotation.config.ConfigMaterialSet", "java.util.Set<org.bukkit.Material>");
        map.put("org.oddlama.vane.annotation.config.ConfigString", "java.lang.String");
        map.put("org.oddlama.vane.annotation.config.ConfigStringList", "java.util.List<java.lang.String>");
        map.put(
            "org.oddlama.vane.annotation.config.ConfigStringListMap",
            "java.util.Map<java.lang.String,java.util.List<java.lang.String>>"
        );
        map.put("org.oddlama.vane.annotation.config.ConfigVersion", "long");
        map.put("org.oddlama.vane.annotation.lang.LangMessage", "org.oddlama.vane.core.lang.TranslatedMessage");
        map.put(
            "org.oddlama.vane.annotation.lang.LangMessageArray",
            "org.oddlama.vane.core.lang.TranslatedMessageArray"
        );
        map.put("org.oddlama.vane.annotation.lang.LangVersion", "long");
        fieldTypeMapping = Collections.unmodifiableMap(map);
    }

    private void verifyType(TypeElement annotation, Element element) {
        var type = element.asType().toString();
        var requiredType = fieldTypeMapping.get(annotation.asType().toString());
        if (requiredType == null) {
            processingEnv
                .getMessager()
                .printMessage(
                    Diagnostic.Kind.ERROR,
                    element.asType().toString() +
                    ": @" +
                    annotation.getSimpleName() +
                    " has no requiredType mapping! This is a bug."
                );
        } else {
            if (!requiredType.equals("<any>") && !requiredType.equals(type)) {
                processingEnv
                    .getMessager()
                    .printMessage(
                        Diagnostic.Kind.ERROR,
                        element.asType().toString() +
                        ": @" +
                        annotation.getSimpleName() +
                        " requires a field of type " +
                        requiredType +
                        " but got " +
                        type
                    );
            }
        }
    }
}

File: vane-annotations/src/main/kotlin/org/oddlama/vane/annotation/processor/ConfigAndLangProcessor.kt
package org.oddlama.vane.annotation.processor

import javax.annotation.processing.AbstractProcessor
import javax.annotation.processing.RoundEnvironment
import javax.lang.model.SourceVersion
import javax.lang.model.element.Element
import javax.lang.model.element.TypeElement
import javax.tools.Diagnostic

/** Placeholder used for annotations whose annotated field may have any type. */
private const val ANY_TYPE = "<any>"

/**
 * Maps each config/lang annotation to the exact Java type string the matching `ConfigField` or
 * `LangField` reader in vane-core expects on the annotated field. This map is also the single
 * source of truth for the annotations this processor claims — see [getSupportedAnnotationTypes].
 */
private val fieldTypeMapping: Map<String, String> = mapOf(
    "org.oddlama.vane.annotation.config.ConfigBoolean" to "boolean",
    "org.oddlama.vane.annotation.config.ConfigDict" to ANY_TYPE,
    "org.oddlama.vane.annotation.config.ConfigDouble" to "double",
    "org.oddlama.vane.annotation.config.ConfigDoubleList" to "java.util.List<java.lang.Double>",
    "org.oddlama.vane.annotation.config.ConfigExtendedMaterial" to "org.oddlama.vane.core.material.ExtendedMaterial",
    "org.oddlama.vane.annotation.config.ConfigInt" to "int",
    "org.oddlama.vane.annotation.config.ConfigIntList" to "java.util.List<java.lang.Integer>",
    "org.oddlama.vane.annotation.config.ConfigItemStack" to "org.bukkit.inventory.ItemStack",
    "org.oddlama.vane.annotation.config.ConfigLong" to "long",
    "org.oddlama.vane.annotation.config.ConfigMaterial" to "org.bukkit.Material",
    "org.oddlama.vane.annotation.config.ConfigMaterialMapMapMap" to
        "java.util.Map<java.lang.String,java.util.Map<java.lang.String,java.util.Map<java.lang.String,org.bukkit.Material>>>",
    "org.oddlama.vane.annotation.config.ConfigMaterialSet" to "java.util.Set<org.bukkit.Material>",
    "org.oddlama.vane.annotation.config.ConfigString" to "java.lang.String",
    "org.oddlama.vane.annotation.config.ConfigStringList" to "java.util.List<java.lang.String>",
    "org.oddlama.vane.annotation.config.ConfigStringListMap" to "java.util.Map<java.lang.String,java.util.List<java.lang.String>>",
    "org.oddlama.vane.annotation.config.ConfigVersion" to "long",
    "org.oddlama.vane.annotation.lang.LangMessage" to "org.oddlama.vane.core.lang.TranslatedMessage",
    "org.oddlama.vane.annotation.lang.LangMessageArray" to "org.oddlama.vane.core.lang.TranslatedMessageArray",
    "org.oddlama.vane.annotation.lang.LangVersion" to "long"
)

/**
 * Annotation processor that validates configuration and language field annotations.
 *
 * It ensures that fields annotated with the config/lang annotations have the
 * exact type the corresponding vane-core field reader will cast them to, per [fieldTypeMapping].
 */
class ConfigAndLangProcessor : AbstractProcessor() {
    /** Claims exactly the annotations [fieldTypeMapping] knows a required type for. */
    override fun getSupportedAnnotationTypes(): Set<String> = fieldTypeMapping.keys

    /**
     * Tracks whatever the running JDK supports instead of pinning a release, so the toolchain
     * can move forward without javac warning about an outdated `@SupportedSourceVersion`.
     */
    override fun getSupportedSourceVersion(): SourceVersion = SourceVersion.latestSupported()

    /**
     * Processes the set of annotation types for this round and validates
     * annotated elements against the expected types.
     *
     * @param annotations The set of annotation types to process.
     * @param roundEnv The processing environment for this round.
     * @return true to indicate annotations are claimed by this processor.
     */
    override fun process(annotations: MutableSet<out TypeElement>, roundEnv: RoundEnvironment): Boolean {
        annotations.forEach { annotation ->
            roundEnv.getElementsAnnotatedWith(annotation).forEach { verifyType(annotation, it) }
        }

        return true
    }

    /**
     * Verifies that the annotated element has the expected type for the
     * given annotation. Emits an error if the mapping is missing or types do not match.
     *
     * @param annotation The annotation type being validated.
     * @param element The element annotated with that annotation.
     */
    private fun verifyType(annotation: TypeElement, element: Element) {
        val type = element.asType().toString()
        val requiredType = fieldTypeMapping[annotation.asType().toString()]
            ?: return processingEnv.messager.printMessage(
                Diagnostic.Kind.ERROR,
                "${element.asType()}: @${annotation.simpleName} has no requiredType mapping! This is a bug."
            )

        if (requiredType != ANY_TYPE && requiredType != type) {
            processingEnv.messager.printMessage(
                Diagnostic.Kind.ERROR,
                "${element.asType()}: @${annotation.simpleName} requires a field of type $requiredType but got $type"
            )
        }
    }
}
